### DISCUSSION TITLE
Model parameters : bookkeeping with 2 different lists

### DISCUSSION THREAD

**Discussion by pariRieck** [scikit-hep/pyhf#1750] [Q&A] [answered]
Dear pyhf developers,

while inspecting model objects, I realised that the list of parameters is provided in 2 different ways:

1. model.config.parameters
2. model.config.par_order

The two lists are ordered differently.

Also there is model.config.par_names() which is similar to model.config.order , but indicates nuisance parameter groups (e.g. staterrors).
With the example sent to you in an e-mail earlier, some confusion was caused due to the different lists, related to model.config.poi_index and which of the parameter lists fits to. I realise now that it is list number 2 (par_order). Still, maybe you could comment on the rationale here.

Best regards
Patrick

**Comment by alexander-held** [comment]
Related issue for clarifying the distinction in the documentation: https://github.com/scikit-hep/pyhf/issues/1747

**Comment by kratsg** [comment, marked as answer]
To give a (naive) rationale, as I suspect this was my doing. I wanted attributes on various models that were "consistent" regardless of the ordering of lists in the JSON. The minor thing about JSON is that you can argue that samples with modifiers `[mod1,mod2]` and `[mod2,mod1]` are equivalent. For us, the order in which modifiers (or samples, or channels) are specified should not impact or change anything about the built model.

In this case, while the `par_order` differs (because the order of parameters is what determines the order of the underlying binning - and similarly for order of samples/channels for groups of bins), that means that the ordering differs between two identical models... but the `sorted` list of parameter names, or channel names, or sample names, should still be identical. Therefore, `model.config.parameters` should be the same while `model.config.par_order` will have changed.

This is also why there are utilities such as `pyhf sort` and `pyhf digest` to help determine if you're looking at equivalent models or not.
